A78-grpc-metrics-wrr-pf-xds.md

```diff
@@ -0,0 +1,154 @@
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.target | required | Indicates the target of the gRPC channel in which PF is used.  (Same as the attribute defined in [A66].) |
+
+The following metrics will be exported:
+
+| Name          | Type  | Unit  | Labels  | Description |
+| ------------- | ----- | ----- | ------- | ----------- |
+| grpc.lb.pick_first.disconnections | Counter | {disconnection} | grpc.target | Number of times the selected subchannel becomes disconnected. |
+| grpc.lb.pick_first.connection_attempts_succeeded | Counter | {attempt} | grpc.target | Number of successful connection attempts. |
+| grpc.lb.pick_first.connection_attempts_failed | Counter | {attempt} | grpc.target | Number of failed connection attempts. |
+
+### XdsClient
+
+The XdsClient component was originally described in [A27].  Note that in
+[A71], we are moving from a single global XdsClient instance to a
+separate global XdsClient instance for each channel target.  The
+proposed metric schema here reflects that change.
+
+XdsClient metrics will have the following labels:
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.target | required | For clients, indicates the target of the gRPC channel in which the XdsClient is used (i.e., the same as the attribute defined in [A66]). For servers, will be the string "#server". |
+| grpc.xds.server | required | The target URI of the xDS server with which the XdsClient is communicating. |
+| grpc.xds.authority | required | The xDS authority.  The value will be "#old" for old-style non-xdstp resource names. |
+| grpc.xds.cache_state | required | Indicates the cache state of an xDS resource.  The value will be one of: <ul><li>"requested": The resource has been requested from the xDS server but has not yet been received.<li>"does_not_exist": The server has indicated that the resource does not exist.<li>"acked": The resource has been received and is valid.<li>"nacked": The resource was received but was not valid.<li>"nacked_but_cached": There is a version of the resource cached, but the most recent update of the resource was invalid.</ul> |
+| grpc.xds.resource_type | required | Indicates an xDS resource type, such as "envoy.config.listener.v3.Listener". |
+
+The following metrics will be exported:
+
+| Name          | Type  | Unit  | Labels  | Description |
+| ------------- | ----- | ----- | ------- | ----------- |
+| grpc.xds_client.connected | Gauge | {bool} | grpc.target, grpc.xds.server | Whether or not the xDS client currently has a working ADS stream to the xDS server.  For a given server, this will be set to 1 when the stream is initially created.  It will be set to 0 when we have a connectivity failure or when the ADS stream fails without seeing a response message, as per [A57].  Once set to 0, it will be reset to 1 when we receive the first response on an ADS stream. |
```

Comment:
Unless the client may connect to multiple xds servers per {target, resource_type}, isn't a string gauge that records the xds server address (or none) better? I think this is closer to how gauge counters are typically used.

Reply:
In the general case, it's entirely possible for a client to talk to multiple servers for a given target, due to federation.